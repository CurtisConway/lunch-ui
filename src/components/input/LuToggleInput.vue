<template>
  <div :class="[defaultClassList, 'lu-toggle-input']">
    <LuText
      tag="label"
      :for="computedId"
      size="subtitle"
    >
      {{ label }}
    </LuText>
    <input
      ref="inputElement"
      :id="computedId"
      type="checkbox"
      :class="bgColorFromColor"
      v-model="valueInterface"
      :checked="valueInterface"
      @change="onChange"
      @focus="onFocus"
      @blur="onBlur"
    >
    <slot :focus="focus"></slot>
  </div>
</template>

<script>
import colors from '../../assets/js/mixins/colors';
import borders from '../../assets/js/mixins/borders';
import spacing from '../../assets/js/mixins/spacing';
import events from '../../assets/js/mixins/events';
import validation from './_validation.mixin';
import inputs from './_inputs.mixin';

export default {
  name: 'LuToggleInput',
  mixins: [colors, borders, spacing, events, inputs, validation],
  props: {
    value: {
      type: Boolean,
      default: () => false,
    },
    validate: {
      type: Boolean,
      default: () => false,
    },
  },
  computed: {
    defaultClassList() {
      return [
        'lu-form-group',
        this.paddingClasses,
        this.marginClasses,
        this.colorClass,
        {
          focus: this.focus,
        },
      ];
    },
  },
};
</script>
