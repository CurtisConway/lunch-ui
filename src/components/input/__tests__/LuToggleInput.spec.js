import Vue from 'vue';
import Router from 'vue-router';

import LuToggleInput from '../LuToggleInput.vue';

import {
  createLocalVue,
  mount,
} from '@vue/test-utils';
import { compileToFunctions } from 'vue-template-compiler'

describe('LuToggleInput.vue', () => {
  let mountFunction;
  let router = Router;
  let localVue = typeof Vue;

  beforeEach(() => {
    router = new Router();
    localVue = createLocalVue();
    localVue.use(Router);

    mountFunction = (options = {}) => {
      return mount(LuToggleInput, {
        localVue,
        router,
        ...options,
      });
    }
  });

  it('should render component and match snapshot', () => {
    expect(mountFunction().html()).toMatchSnapshot();
  });

  it('should render component with a value and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        value: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with a label prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        label: 'test',
      },
    }).html()).toMatchSnapshot();
  });

  it('should emit a focus and blur event', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        value: false,
      },
    });

    const focus = jest.fn();
    const blur = jest.fn();
    const inputElement = wrapper.vm.$refs.inputElement;
    wrapper.vm.$on('focus', focus);
    wrapper.vm.$on('blur', blur);
    inputElement.focus();
    inputElement.blur();
    inputElement.focus();
    inputElement.blur();

    expect(focus.mock.calls).toHaveLength(2);
    expect(blur.mock.calls).toHaveLength(2);
  });

  it('should emit an change event', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        value: false,
      },
    });

    const event = jest.fn();
    wrapper.vm.$on('change', event);
    const input = wrapper.find('input');
    await input.setChecked(true);
    await input.setChecked(false);

    expect(event.mock.calls).toHaveLength(2);
  });
});
