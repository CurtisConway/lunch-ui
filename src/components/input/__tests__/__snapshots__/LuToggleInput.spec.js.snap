// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuToggleInput.vue should render component and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-toggle-input"><label class="lu-text text-subtitle text-left" for="input-1">

  </label> <input id="input-1" type="checkbox" class="bg-black"> </div>
`;

exports[`LuToggleInput.vue should render component with a value and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-toggle-input"><label class="lu-text text-subtitle text-left" for="testInput">

  </label> <input id="testInput" type="checkbox" class="bg-black"> </div>
`;

exports[`LuToggleInput.vue should render with a label prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-toggle-input"><label class="lu-text text-subtitle text-left" for="testInput">
    test
  </label> <input id="testInput" type="checkbox" class="bg-black"> </div>
`;
